File: 2059-unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        unordered_set<char> lst;
        //for optimized calculation we add each character once only
        for(char c: s){
            lst.insert(c);
        }

        int ans = 0;
        //check first and last occurance of each char
        for(char ltr: lst){
            int i = -1, j = 0;

            for(int k=0; k<s.size(); k++){
                if(s[k]==ltr){
                    if(i == -1){
                        i = k;
                    }
                    j = k;
                }
            }

            unordered_set<char> btwn;
            for(int p=i+1; p<j; p++){
                btwn.insert(s[p]);
            }

            ans += btwn.size();
        }

        return ans;
    }
};

File: 2059-unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.java
import java.util.HashSet;

class Solution {
    public int countPalindromicSubsequence(String s) {
        HashSet<Character> uniqueChars = new HashSet<>();
        
        // Store all unique characters in the string
        for (char c : s.toCharArray()) {
            uniqueChars.add(c);
        }
        
        int result = 0;
        
        // Process each unique character
        for (char ltr : uniqueChars) {
            int first = -1, last = 0;
            
            // Find the first and last occurrences of the current character
            for (int k = 0; k < s.length(); k++) {
                if (s.charAt(k) == ltr) {
                    if (first == -1) {
                        first = k;
                    }
                    last = k;
                }
            }
            
            // Count unique characters between the first and last occurrences
            HashSet<Character> between = new HashSet<>();
            for (int p = first + 1; p < last; p++) {
                between.add(s.charAt(p));
            }
            
            result += between.size();
        }
        
        return result;
    }
}
